.envrc

```diff
@@ -0,0 +1 @@
+use_nix
```

Comment:
What about `direnv deny`? Also works with `direnv disallow`, `direnv block`, and `direnv revoke`

Reply:
I honestly don't understand your point. Committing the `.envrc` file is the *point* of direnv, to make the developer experience smoother. I love it when projects do that.  The shell is additive and not pure, so it will take the outside environment into account.

Reply:
> This isn’t the experience a user wants.

As a user, I expect that when I install a tool (direnv) and a repo supports it (nixfmt), then the tool should be used. Don't want direnv? Don't install it, then.

> I need different environment variables & scripts specific to me

Great, direnv has support for that: `$XDG_CONFIG_HOME/direnv/direnvrc` and `$XDG_CONFIG_HOME/direnv/lib/*.sh`. We could also do [this](https://github.com/direnv/direnv/issues/556#issuecomment-576197199) but that has security implications.

To be honest, I have never heard from any developer who has direnv installed that they don't want to use it (why did they install it then?). To me it's as simple as:
- Only editing the readme once? Ignore the warning
- Doing something in the repo often? `direnv deny` it once. It's quick and easy
- Doing something in the repo often but you are like me? Ignore the warning every time

Reply:
Alternative option: Source a `.envrc.local` (or another name) from `.envrc`, for user-local setup. This could also be given a mechanism to disable the default `use nix`-invocation. Very similar, but also a bit different, in that it offers both a) default-setup in `.envrc` out-of-the-box (i.e., when cloning and `cd`ing into the repo), as well as b) local setup/overrides.

Reply:
See the comment I linked. I was going to suggest the same but the security concerns are unfortunately legitimate :/

Reply:
> See the comment I linked. I was going to suggest the same but the security concerns are unfortunately legitimate :/

Oh, your comment wasn't there yet when I started writing mine :sweat_smile: Good point, didn't know about that, ty! :)

Reply:
I will bring this to discussion in the next Formatting team meeting, as part of the planned code review for this pull request. Thank you for your input.

Reply:
Also, please don't block this PR on that, I think everyone can handle whatever outcome we end up with :see_no_evil: Thank you for your work!! :)

Reply:
> To be honest, I have never heard from any developer who has direnv installed that they don't want to use it (why did they install it then?). To me it's as simple as:
>    - Only editing the readme once? Ignore the warning
>   - Doing something in the repo often? direnv deny it once. It's quick and easy
>    - Doing something in the repo often but you are like me? Ignore the warning every time

@dasJ Late to the party, but I just wanted to say I strongly discourage adding `.envrc` to a git repo (or any other VCS), and it's not about having `direnv` installed and not wanting to use it btw :) Here are some counter arguments:

- I want to have my own nix shell and env vars when working on this repo (or any other), for which I would like to be in control of my own `.envrc` file.
   * Specifically on this project, what if I want to use a different [Haskell LSP server](https://github.com/NixOS/nixfmt/blob/2d79e0fdb5a83253b4598dd2e33810cf08754ffb/default.nix#L117) version? or a different Cabal version? Or if I need some other package in this environment that the project assumes it's globally available?
- I want to be able to source other local files relevant to my environment on that `.envrc`, e.g. I use fish shell and it requires some tweaks on certain repos, other folks use zsh or bash, etc.

Having this file committed to the repo removes this flexibility, but it's also easier to argue in its favor for an open source public repo. For private repos there's even more counter arguments, for instance, we have this at work:

- Need to set AWS secrets on a daily basis (lasts 24 hours).
- Need to set specific secrets that last a year (can be sourced from a local file).
- Still need specific Nix shells at times (every project is different).
- Need project-specific env vars that need to change depending on what we are testing and on which environment.

### What I suggest instead

Let the users decide what they want available in their environments while still leveraging `direnv`, be a good citizen and add `.envrc` to your global `.gitignore` or add it to the project's `.gitignore` to avoid anyone adding it back.

Reply:
> Or if I need some other package in this environment that the project assumes it's globally available?

a good project shell would assume no such things 

Regarding credentials: these should not be set in .envrc, but project shell should provide tools (scripts, commands) to manage these 

> Still need specific Nix shells at times (every project is different).

These can be entered ad-hoc with nix shell 

> Need project-specific env vars that need to change depending on what we are testing and on which environment.

A command (provided by default dev shell) should be provided to switch between envs 

.envrc IMHO should never be edited by devs, it should simply setup all tools needed to work on given project. If customisation of env vars is needed this could be loaded from optional, gitignored file(s)

Reply:
> a good project shell would assume no such things

This project assumes I need a specific HLS version

> Regarding credentials: these should not be set in .envrc, but project shell should provide tools (scripts, commands) to manage these

Agree to disagree. The methods vary across different projects when working on a big organization. A lot of flows are very complex and unfortunately can't be automated, e.g. sign in via SSO behind a VPN, approving the sign in attempt via Okta / Duo, etc. It barely works on Linux, let alone trying to automate these things.

It's much easier on open source public projects, I'll give you that, but private organizations are a completely different story.

> These can be entered ad-hoc with nix shell\develop

It doesn't make it easier to pin specific pkgs, add overlays, etc.

> A command (provided by default dev shell) should be provided to switch between envs

This isn't bad actually, might work well in some cases, will give it a try :) Certainly not in cases where switching envs requires going through the complex authorization flows, though.

> If customisation of env vars is needed this could be loaded from optional, gitignored file(s)

How would you do this if you don't control the content of `.envrc`?

Reply:
We've had this topic at least thrice already. Not sure having this discussion yet another time is a productive use of anybody's time